<a href="https://colab.research.google.com/github/LinhThaoPham/DailyHomeWork/blob/main/btvn1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Restaurant

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **Import data and libraries**

Sử dụng restaurant dataset từ 'data\\restaurant.csv'

Bộ dữ liệu chứa thông tin tiền tip từ nhiều khách hàng khác nhau, bao gồm nữ và nam, người hút thuốc và không hút thuốc, từ các ngày khác nhau, bữa tối hoặc trưa, và với kích thước bàn ăn khác nhau.

In [7]:
import pandas as pd

# Load restaurant.csv file from data folder to variable "df":
df = pd.read_csv('/content/drive/MyDrive/data/restaurant.csv')


In [9]:
# In ra 5 hàng đầu tiên
df.head(5)


total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## **Overview**

### 1. Write **one line of code** to show the **first 3 rows** of the data

**Output should look like this**

|    |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|---:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
|  0 |        16.99 |  1.01 | Female | No       | Sun   | Dinner |      2 |
|  1 |        10.34 |  1.66 | Male   | No       | Sun   | Dinner |      3 |
|  2 |        21.01 |  3.5  | Male   | No       | Sun   | Dinner |      3 |

In [10]:
# Your answer/code here:
df.head(3)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3

### 2. Write **one line of code** to Show the **last 2 rows** of the data

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 242 |        17.82 |  1.75 | Male   | No       | Sat   | Dinner |      2 |
| 243 |        18.78 |  3    | Female | No       | Thur  | Dinner |      2 |

In [11]:
# Your answer/code here:
df.tail(2)

total_bill   tip     sex smoker   day    time  size
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

### 3. What syntax is right to get random rows in dataframe?

-  A df.shuffle()

- B df.random()

- C df.sample()

- D df.mix()

In [18]:
# Your answer/code here:
df.sample()
# Viết thêm ý nghĩa các hàm khác


total_bill   tip   sex smoker   day   time  size
88       24.71  5.85  Male     No  Thur  Lunch     2

### 4. Write one line of code to show a **descriptive statistics summary** of the dataset.

**Output should look like this**

|       |   total_bill |       tip |      size |
|:------|-------------:|----------:|----------:|
| count |    244       | 244       | 244       |
| mean  |     19.7859  |   2.99828 |   2.56967 |
| std   |      8.90241 |   1.38364 |   0.9511  |
| min   |      3.07    |   1       |   1       |
| 25%   |     13.3475  |   2       |   2       |
| 50%   |     17.795   |   2.9     |   2       |
| 75%   |     24.1275  |   3.5625  |   3       |
| max   |     50.81    |  10       |   6       |

In [13]:
# Your answer/code here:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

## **EDA - Exploratory Data Analysis**

### 5. Write one line of code to find out the **minimum amount of tip**.

**Output should be a number**

In [19]:
# Your answer/code here:
df['tip'].min()

1.0

### 6. Write one line of code to find out the **maximum amount of tip**.

**Output should be a number**

In [20]:
# Your answer/code here:
df['tip'].max()

10.0

### 7. On which days of the week, the restaurant is open?

- A. Mon, Tue, Wed, Thu

- B. Thu, Fri, Sat, Sun

- C. Mon, Wed, Fri

- D. Open all days in week

In [22]:
# Your answer/code here:
df['day'].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

### 8. Write one line of code to get the list of **unique values** of the column ```time```.

**Output should look like this**

array(['Dinner', 'Lunch'], dtype=object)

In [24]:
# Your answer/code here:
df['time'].unique()

array(['Dinner', 'Lunch'], dtype=object)

### 9. Write one line of code to get the range of table size.

Range means the **difference between the max and the min**.

**Output should be a number**

In [28]:
# Your answer/code here:
df['size'].max() - df['size'].min()

5

### 10. How many men/women paid for the dish?

- A Male: 157, Female: 89

- B Male: 85, Female: 156

- C Male: 157, Female: 87

- D Male: 142, Female: 15

In [30]:
# Your answer/code here:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [35]:
df.groupby('sex').count()

total_bill  tip  smoker  day  time  size
sex                                             
Female          87   87      87   87    87    87
Male           157  157     157  157   157   157

### 11. What is the percentage of number of smoker table over the total?

- A 38%

- B 40%

- C 2%

- D 62%

In [49]:
# Your answer/code here:
(df['smoker'] == 'Yes').sum()/ df['smoker'].count() * 100

np.float64(38.114754098360656)

### 12. What is the percentage of revenue from the smoker tables over the total?

- A 38%

- B 40%

- C 2%

- D 62%

In [50]:
# Your answer/code here:
df[df['smoker'] == 'Yes']['total_bill'].sum() / df['total_bill'].sum() * 100

np.float64(39.98409203421041)

### 14. Between smoker and non-smoker, who tips more according the percentage of the total tip on total bill.

- A Non-smoker

- B Smoker

In [53]:
# Your answer/code here:
df.groupby('smoker')['tip'].sum() / df.groupby('smoker')['total_bill'].sum() * 100

smoker
No     15.592094
Yes    14.495374
dtype: float64

### 15. Write one line of code to filter out the **"smoker"** tables on **Sunday**.

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 164 |        17.51 |  3    | Female | Yes      | Sun   | Dinner |      2 |
| 172 |         7.25 |  5.15 | Male   | Yes      | Sun   | Dinner |      2 |
| 173 |        31.85 |  3.18 | Male   | Yes      | Sun   | Dinner |      2 |
| 174 |        16.82 |  4    | Male   | Yes      | Sun   | Dinner |      2 |
| 175 |        32.9  |  3.11 | Male   | Yes      | Sun   | Dinner |      2 |
| 176 |        17.89 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 177 |        14.48 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 178 |         9.6  |  4    | Female | Yes      | Sun   | Dinner |      2 |
| 179 |        34.63 |  3.55 | Male   | Yes      | Sun   | Dinner |      2 |
| 180 |        34.65 |  3.68 | Male   | Yes      | Sun   | Dinner |      4 |
| 181 |        23.33 |  5.65 | Male   | Yes      | Sun   | Dinner |      2 |
| 182 |        45.35 |  3.5  | Male   | Yes      | Sun   | Dinner |      3 |
| 183 |        23.17 |  6.5  | Male   | Yes      | Sun   | Dinner |      4 |
| 184 |        40.55 |  3    | Male   | Yes      | Sun   | Dinner |      2 |
| 186 |        20.9  |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 187 |        30.46 |  2    | Male   | Yes      | Sun   | Dinner |      5 |
| 188 |        18.15 |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 189 |        23.1  |  4    | Male   | Yes      | Sun   | Dinner |      3 |
| 190 |        15.69 |  1.5  | Male   | Yes      | Sun   | Dinner |      2 |

In [59]:
# Your answer/code here:
df[(df['smoker'] == 'Yes') & (df['day'] == 'Sun')]


total_bill   tip     sex smoker  day    time  size
164       17.51  3.00  Female    Yes  Sun  Dinner     2
172        7.25  5.15    Male    Yes  Sun  Dinner     2
173       31.85  3.18    Male    Yes  Sun  Dinner     2
174       16.82  4.00    Male    Yes  Sun  Dinner     2
175       32.90  3.11    Male    Yes  Sun  Dinner     2
176       17.89  2.00    Male    Yes  Sun  Dinner     2
177       14.48  2.00    Male    Yes  Sun  Dinner     2
178        9.60  4.00  Female    Yes  Sun  Dinner     2
179       34.63  3.55    Male    Yes  Sun  Dinner     2
180       34.65  3.68    Male    Yes  Sun  Dinner     4
181       23.33  5.65    Male    Yes  Sun  Dinner     2
182       45.35  3.50    Male    Yes  Sun  Dinner     3
183       23.17  6.50    Male    Yes  Sun  Dinner     4
184       40.55  3.00    Male    Yes  Sun  Dinner     2
186       20.90  3.50  Female    Yes  Sun  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
188       18.15  3.50  Female    Yes  Sun  Dinner     3
189       23.10  4.00    Male    Yes  Sun  Dinner     3
190       15.69  1.50    Male    Yes  Sun  Dinner     2

### 16. How many tables belong to the smoker, at lunch?

- A 20

- B 21

- C 22

- D 23

In [63]:
# Your answer/code here:
df[(df['smoker'] == 'Yes') & (df['time'] == 'Lunch')].count()

total_bill    23
tip           23
sex           23
smoker        23
day           23
time          23
size          23
dtype: int64

### 17. Write one line of code to compute the **average tip** amount of men who smoke and dine on **Friday's dinner**.


**Output should be a number**

In [64]:
# Your answer/code here:
df[(df['sex'] == 'Male') & (df['smoker'] == 'Yes') & (df['day'] == 'Fri') & (df['time'] == 'Dinner')]['tip'].mean()

np.float64(3.246)

### 18. Write one line of code to compute the **sum** of total bills of the **male** customers who dine in any day that's **not Friday**.

**Output should be a number**

In [66]:
# Your answer/code here:
df[(df['sex'] == 'Male') & (df['day'] == 'Fri')]['total_bill'].sum()

np.float64(198.57)

### 19. Write one line of code to filter the tables with tip amount **larger** than the **average** tip amount.

In [67]:
# Your answer/code here:
df[df['tip'] > df['tip'].mean()]

total_bill   tip     sex smoker   day    time  size
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
5         25.29  4.71    Male     No   Sun  Dinner     4
7         26.88  3.12    Male     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
232       11.61  3.39    Male     No   Sat  Dinner     2
234       15.53  3.00    Male    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
239       29.03  5.92    Male     No   Sat  Dinner     3
243       18.78  3.00  Female     No  Thur  Dinner     2

[121 rows x 7 columns]

### 20

- Task 1: load `time_hours.xlsx` table to `df_th` variable:

In [68]:
df_th = pd.read_excel('/content/drive/MyDrive/data/time_hours.xlsx')

# show time_hours table:
df_th

time    hour
0  Dinner  19-21h
1   Lunch  10-12h

- Task 2: Get `hour` information for `df` dataframe from `df_th` by using Using `time` field (`pd.merge()`):

In [70]:
# your code here:
df = pd.merge(
    df,
    df_th,
    how='left',
    on='time'
)

In [72]:
# show random sample 5 rows:
df.sample(5)

total_bill   tip     sex smoker   day    time  size  hour_x  hour_y
171       15.81  3.16    Male    Yes   Sat  Dinner     2  19-21h  19-21h
189       23.10  4.00    Male    Yes   Sun  Dinner     3  19-21h  19-21h
79        17.29  2.71    Male     No  Thur   Lunch     2  10-12h  10-12h
190       15.69  1.50    Male    Yes   Sun  Dinner     2  19-21h  19-21h
82        10.07  1.83  Female     No  Thur   Lunch     1  10-12h  10-12h